File: ruby/016_MinimumCoins.rb
# 016 - Minimum Coins（★3）
# 工夫した全探索

N = gets.to_i
A, B, C = gets.split.map(&:to_i).sort.reverse

# 制約よりコインの枚数は最大でも9999。
# 最小枚数を求めるため、ループはans以下で繰り返す
# 外側ループはAよりBの方が速い
ans = 9999
[N / B, ans].min.downto(0) do |b|
  [(N - B * b) / A, ans - b].min.downto(0) do |a|
    c, remain = (N - (A * a + B * b)).divmod(C)
    sum = a + b + c
    if remain.zero?
      ans = [ans, sum].min
      break
    # ansを超える場合はループを打ち切る
    # sum > ansだとTLEする（sum + 1 >= ansの方が速い）
    elsif sum + 1 >= ans
      break
    end
  end
end

puts ans

File: kotlin/016_MinimumCoins.kt
/**
 * 016 - Minimum Coins（★3）
 * 工夫した全探索
 */

fun main() {
  val MAX = 9_999
  val N = readLine()!!.toInt()
  val (A, B, C) = readLine()!!.split(" ").map(String::toInt).sorted().reversed()

  var ans = MAX
  for (a in Math.min(MAX, N / A) downTo 0) {
    for (b in Math.min(MAX - a, (N - A * a) / B) downTo 0) {
      val c = (N - A * a - B * b) / C
      val sum = a + b + c
      if (A * a + B * b + C * c == N) {
        ans = Math.min(ans, sum)
        break
      } else if (ans < sum) break
    }
  }
  println(ans)
}
